Added general compatibility guidelines, but with focus on logging (#984)
Azure/azure-sdk: docs/general/compatibility.md

---
title: "General Guidelines: Compatibility"
keywords: guidelines
permalink: general_compatibility.html
folder: general
sidebar: general_sidebar
---

This section describes guidelines related to compatibility. In particular, it describes high-level compatibility principles, and detailed rules for reasoning about API and behavioral changes introduced as we evolve our libraries.

## Compatibility Principles
One of the main general principles of the Azure SDK is that it is **Dependable**, and one of the main aspects of dependability is a high level of compatibility. 

> In the context of these guideline, compatibility means that software written using one version of our libraries should function correctly when upgraded to a newer version of the same library, as long as it calls the same version of the service REST API.

Reviewer: "same version of the service REST API" is somewhat ambiguous in this context - does it imply that the software uses the same API version after the upgrade as before (I assume this is not the intended meaning), or does it mean that within my application, all calls use the same API version (I assume that is what it means)?

Author: Same version of REST APIs as it was calling before the update, i.e. if you specify the service version explicitly when instantiating the client.

Reviewer: Oh, ok. I'm not sure if the casual reader of the doc will grok this.

Author: I will clarify

> This is to ensure that developers using the Azure SDK can upgrade to new versions of our libraries without immediattely being subjected to a large cost of fixing resulting breaks. 

Reviewer 2: typo: immediately

This is good for our customers, as it lowers their costs, and it's good for us, as it helps in the adoption of new Azure features, i.e. a win-win.

Having said that, you can always imagine code, even if contrived, that will break when upgraded to a new version of a library that simply adds functionality, fixes bugs, or runs faster. 
Also, there are some incompatible changes that are worth doing, e.g. changes required to fix security bugs. 
And so given that absolute compatibility is not possible in practice, we want to be best-in-class, which as a rule of thumb means being on par with base libraries of the corresponding language.

> Principle #1: Azure SDK libraries must be as compatible or better than the base libraries of their language. For example, Azure SDK for .NET libraries as compatible as the .NET Base Class Libraries.

Reviewer: I like this principle. It has enough flex to not be brittle.

> Principle #1: Azure SDK libraries must be as compatible or better than the base libraries of their language. For example, Azure SDK for .NET libraries as compatible as the .NET Base Class Libraries.

Reviewer 3: ```suggestion
Principle #1: Azure SDK libraries must be as compatible or better than the base libraries of their language. For example, Azure SDK for .NET libraries must be as compatible as the .NET Base Class Libraries.
```

> Principle #1: Azure SDK libraries must be as compatible or better than the base libraries of their language. For example, Azure SDK for .NET libraries as as compatible as the .NET Base Class Libraries.

Reviewer 2: typo: "as as"

### Azure SDK Compatibility Reviews
Azure SDK guidelines define details of what's considered a compatible change in each language ecosystem. All changes that are not explicitly designated as compatible, need to be reviewed and approved by the Azure SDK Architecture Board. 

### API Changes

> It might seem that additions to the APIs would be non-breaking, and removals would be breaking. Unfortunately, it's not as simple. Languages differ quite significantly in terms of what API changes would result in breaks in the caller. In particular, there are very significant differences between dynamic and static languages. Therefore, you should refer to language specific guidelines for details of what constitutes an acceptable API change.

Reviewer: Would it make sense to give the example of adding a method to an interface for .NET? I don't think everybody has internalized this - I have definitely come across the argument that it only is relevant for interfaces that the user may implement, and for interfaces that are both defined by and only implemented by a client library, it "cannot" break clients.

Author: How does this comment relate to the text above? The text above does not talk about adding APIs to interfaces.

Reviewer 2: Johan is suggesting an example where adding a method or field/propety would break since most people would consider this a non-breaking change. Since interfaces are supposed to immutable, Johan is suggesting an interface example that mutates the interface by adding something is breaking. Johan: This would be a good example except that the .NET library tries hard to avoid using interfaces at all. 

But, perhaps an example of a service returning a new field which is not round-trippable. Or an example of an API going from non-paging to paging?

> It might seem that additions to the APIs would be non-breaking, and removals would be breaking. Unfortunatelly, it's not as simple. Languages differ quite significanly in terms of what API changes would result in breaks in the caller. In particular, there are very significant differences between dynamic and and static languages. Therefore, you should refer to language specific guidelines for details of what constitutes an acceptable API change.

Reviewer 2: typo: "and and"

> It might seem that additions to the APIs would be non-breaking, and removals would be breaking. Unfortunatelly, it's not as simple. Languages differ quite significanly in terms of what API changes would result in breaks in the caller. In particular, there are very significant differences between dynamic and and static languages. Therefore, you should refer to language specific guidelines for details of what constitutes an acceptable API change.

Reviewer 3: ```suggestion
It might seem that additions to the APIs would be non-breaking, and removals would be breaking. Unfortunately, it's not as simple at this. Languages differ quite significantly in terms of what API changes would result in breaks in the caller. In particular, there are very significant differences between dynamic and static languages. Therefore, you should refer to language-specific guidelines for details of what constitutes an acceptable API change.
```

Behavioral compatibility is more language independent, and so it's described below. 

### Behavioral Changes

TBD

#### Logging Changes

> Changes to the output of logs generated by a library are treated separately from other behavioral change, and they are described in this section.

Reviewer 3: One thing we spoke about previously was having some degree of 'amnesty period' where we are not beholden to these requirements until we feel we have gotten our logging output reviewed and in order. I am not suggesting that we document this, but I think we should discuss the timeframe in which we have to make these improvements, if necessary, before this requirement comes into effect.

Author: I wonder if we need to add it to the guidelines, or just exempt libraries if they need to do it.

Reviewer 3: exemption is fine by me. I would like to have a more clear vision on when we should aim to be in line with this guidance, because I suspect it will take a good amount of dedicated time to evaluate and consider all log output. Similarly, did you want to have any commentary about testing for log output, or are we going to try to prevent it as part of PR review only?

> Changes to the output of logs generated by a library are treated separatelly from other behavioral change, and they are described in this sectio.

Reviewer 2: typo: separately

Many customers build scripts and tools that automatically analyze logs. Just like we don't want software calling our APIs to break when the software is upgraded to new versions of our libraries, we don't want log analysis tools and scripts to break.

> Logs can be seen as a sequence of entries, and entries can be seen as a sequence of tokens that can be parsed by tools. Entry tokens can be further devided into schema related tokens (i.e. tokens that don't vary between instances of entries), and data tokens (tokens that can, and often do, vary between instances of entries, e.g. the time when an entry was logged).

Reviewer 2: typo: devided s/b divided

Changes allowed in a new major or minor version of a package (not a patch version):

> - Additional log entries, e.g. new version logs request details on every retry, while the old version logged only once per retry.

Reviewer: I'm on the fence on this one. I think it is fine to not change log entries. But I'm wondering if we want to allow new log entries.

I don't know any existing libraries in the Python ecosystem that makes this promise.

Reviewer: Does it make sense to distinguish between levels here? I can see this being true for warning and error level messages. I'm on the fence for information. And I don't think we need the restriction for debug.

Author: What is the restriction you are talking about? The text above your comment allows adding log entries; it does not seem to restring anything.

Reviewer 2: If the customer has the ability to filter logs then we can't control what is or is not present in the final log. I think we want to commit to not changing the format of particular log entries. Since sounds nice but I think we might want to modify over time. For example, I could us adding CPU usage or memory usage or some additional metrics over time to some log entries (this is what the CSS team always wanted me to do).

- Additional entry types (entries with a new schema).

All other changes allowed only when:

- Approved by the Azure SDK Architecture Board, and 

> - User must to opt into the new behavior by specifying a new API version or a client option.

Reviewer 3: ```suggestion
- User must opt into the new behavior by specifying a new API version or a client option.
```

> - User must to opt-into the new behavior by specifying a new API version or a client option.

Reviewer 2: grammar: "to opt-in" s/b User must opt into (no hyphen) the new behavior...
